serverless/serverless
# Exclude dev dependency .bin executables

## What did you implement:

Closes #4359 



## How did you implement it:

The `bin` properly `package.json` files of dev deps are read to check the name of the executable(s). If they exist, they are added to the glob exclude patterns.

Existing tests were updated w/ appropriate `readFileAsync` call counts & resolve values. 2 new tests were added handling executable exclusion.

NOTE: `readFileAsync` is being used to read `package.json` files & loading them into memory to access the `.bin` property which is not ideal, but this seems to be the safest way to know the name of the executables. For ex. a lib may be named `lumo-cljs`, but the executable specified in `package.json` could be named `lumo` w/o reading the file, there would be no way to know that. Also, there could be multiple executables include in the [bin](https://docs.npmjs.com/files/package.json#bin) section such as

package.json

```
{
  "bin": {
    "executable1", "executable1.js",
    "executable2", "executable2.js"
  }
}
```

Any feedback would be helpful as this is my first contrib to this project!

## How can we verify it:

Run `yarn test` or pull these changes and try installing as a package locally then running `sls package`

Pull these changes and create a new project, `yarn init` and add a few dev & prod packages:

package.json

```
{
  "name": "sample-project",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "sls": "sls"
  },
  "dependencies": {
    "lodash": "^4.17.4",
    "lokka": "^1.7.0",
    "lokka-transport-http": "^1.6.1"
  },
  "devDependencies": {
    "serverless": "file:path-to-local-serverless",
    "lumo-cljs": "^1.8.0-beta",
  }
}
```

Run

```
$ sls create -t aws-nodejs
$ sls package
$ unzip .serverless/aws-nodejs.zip -lf | grep lumo
```

Should not print anything

## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
